# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Шаг 3. Исследуйте данные и ответьте на вопросы

**Процент должников среди всех заемщиков:**

In [22]:
(data['debt'].mean()*100).round(1)

8.1

### Есть ли зависимость между количеством детей и возвратом кредита в срок?

**Количество заемщиков по количеству детей:**

In [23]:
# количество записей для каждого количества детей
data.groupby('children')['debt'].count()

children
0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: debt, dtype: int64

**Количество должников по количеству детей:**

In [24]:
# количество должников для каждого количества детей
data.groupby('children')['debt'].sum()

children
0    1063
1     444
2     194
3      27
4       4
5       0
Name: debt, dtype: int64

**Процент должников от всех заемщиков по количеству детей:**

In [25]:
# процент должников от общего количества заемщиков
(data.groupby('children')['debt'].mean()*100).round(1)

children
0    7.5
1    9.2
2    9.5
3    8.2
4    9.8
5    0.0
Name: debt, dtype: float64

**Вывод:**

Зависимость между количеством детей и возвратом кредита в срок **есть** \
Наибольшая разница наблюдается между 0 детей и 1 ребенком. При наличии 5 детей видим аномальное значение 0%, вероятно, это связано с небольшим количеством таких людей и непрезентативностью выборки. \
Наименьший процент должников: 0 детей, 7.5% \
Наибольший процент должников: 4 ребенка, 9.8% \
Отсутствие детей дает 7.5% должников, что на 0.6 процентных пункта ниже среднего показателя (8.1%)
Наличие хотя бы одного ребенка дает 9.2% должников, что на 0.9 п.п. выше среднего. При повышении количества детей процент должников растет в среднем на 0.2 п.п. с каждым ребенком.

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

**Все типы семейного положения:**

In [26]:
# все значения семейного положения
data['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'вдовец / вдова',
       'в разводе', 'Не женат / не замужем'], dtype=object)

**Количество заемщиков по семейному положению:**

In [27]:
# количество записей для каждого статуса
data.groupby('family_status')['debt'].count()

family_status
Не женат / не замужем     2796
в разводе                 1189
вдовец / вдова             951
гражданский брак          4134
женат / замужем          12261
Name: debt, dtype: int64

**Количество должников по семейному положению:**

In [28]:
# количество должников для каждого статуса
data.groupby('family_status')['debt'].sum()

family_status
Не женат / не замужем    273
в разводе                 84
вдовец / вдова            63
гражданский брак         385
женат / замужем          927
Name: debt, dtype: int64

**Процент должников от общего количества заемщиков по семейному положению:**

In [29]:
# процент должников для каждого статуса
(data.groupby('family_status')['debt'].mean().sort_values(ascending=False)*100).round(1)

family_status
Не женат / не замужем    9.8
гражданский брак         9.3
женат / замужем          7.6
в разводе                7.1
вдовец / вдова           6.6
Name: debt, dtype: float64

**Рассмотрим связь возраста и семейного положения:**

In [30]:
# медианный возраст заемщика для каждого статуса
data.groupby('family_status')['dob_years'].median().sort_values()

family_status
Не женат / не замужем    36.0
гражданский брак         41.0
женат / замужем          43.0
в разводе                46.0
вдовец / вдова           58.0
Name: dob_years, dtype: float64

**Рассмотрим связь дохода и семейного положения:**

In [31]:
# медианный доход на каждый статус
data.groupby('family_status')['total_income'].median().sort_values(ascending=False)

family_status
в разводе                143503.0
женат / замужем          142916.0
Не женат / не замужем    142594.0
гражданский брак         142594.0
вдовец / вдова           127310.0
Name: total_income, dtype: float64

**Вывод:** 

*Корреляция* между семейным положением и возвратом кредита в срок **есть**, но *зависимость* **не определена**. \
Наименьший процент должников: вдовец / вдова, 6.6% \
Наибольший процент должников: не женат / не замужем, 9.8% \
Оба этих семейных статуса обозначают человека не состоящего в долгосрочных отношениях, тем не менее, значения возврата кредита сильно отличаются. \
При анализе медианного **возраста** наблюдается **обратная корреляция** с семейным положением – возраст растет, а процент должников падает. \
При анализе медианного **дохода** корреляции с семейным положением **нет**.\
Таким образом, можно сделать вывод, что чем более человек взрослый, и чем дальше он прошел по семейному положению вверх (холост, в браке, в разводе и тд.), тем с большей вероятностью он вернет кредит в срок.

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

**Все категории дохода:**

In [32]:
# все категории дохода
sorted(data['total_income_category'].unique())

['A', 'B', 'C', 'D', 'E']

**Все заемщики по доходу:**

In [33]:
# кол-во записей на каждую категорию
data.groupby('total_income_category')['debt'].count()

total_income_category
A       25
B     5014
C    15921
D      349
E       22
Name: debt, dtype: int64

**Все должники по доходу:**

In [34]:
# кол-во должников на каждую категорию
data.groupby('total_income_category')['debt'].sum()

total_income_category
A       2
B     354
C    1353
D      21
E       2
Name: debt, dtype: int64

**Процент должников от всех заемщиков по доходу:**

In [35]:
(data.groupby('total_income_category')['debt'].mean()*100).round(1)

total_income_category
A    8.0
B    7.1
C    8.5
D    6.0
E    9.1
Name: debt, dtype: float64

**Вывод:** 

Категории A, D и E составляют всего 25, 349 и 22 записи соответственно, что в сравнении с категорией C (15921 записей) не может быть репрезентативной выборкой. \
Анализируя оставшиеся категории B и C видим снижение процента должников с 8.5 (выше среднего на 0.4 п.п.) до 7.1 (ниже среднего на 1 п.п.) с ростом дохода.\
Таким образом можно прийти к выводу, что зависимость между уровнем дохода и возвратом кредита в срок **есть**

### Как разные цели кредита влияют на его возврат в срок?

**Все цели кредита:**

In [36]:
# все цели кредита
data['purpose_category'].unique()

array(['операции с недвижимостью', 'операции с автомобилем',
       'получение образования', 'проведение свадьбы'], dtype=object)

**Количество заемщиков по целям кредита:**

In [37]:
# кол-во записей для каждой цели
data.groupby('purpose_category')['debt'].count()

purpose_category
операции с автомобилем       4279
операции с недвижимостью    10751
получение образования        3988
проведение свадьбы           2313
Name: debt, dtype: int64

**Количество должников по целям:**

In [38]:
# кол-во долгов для каждой цели
data.groupby('purpose_category')['debt'].sum()

purpose_category
операции с автомобилем      400
операции с недвижимостью    780
получение образования       369
проведение свадьбы          183
Name: debt, dtype: int64

**Процент должников от всех заемщиков по целям:**

In [39]:
(data.groupby('purpose_category')['debt'].mean().sort_values(ascending=False)*100).round(1)

purpose_category
операции с автомобилем      9.3
получение образования       9.3
проведение свадьбы          7.9
операции с недвижимостью    7.3
Name: debt, dtype: float64

**Вывод:** 

Зависимость возврата кредита от его цели **есть**\
Кредит на автомобиль или образование дают по 9.3% должников (выше среднего на 1.2 п.п.), в то время как проведение свадьбы 7.9% (ниже ср. на 0.2 п.п.), операции с недвижимостью 7.3% (ниже ср. на 0.8 п.п.).\
Наиболее ответственные – приобретатели недвижимости.\
Наименее ответственные - кредиторы на образование и автомобили.

### Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 

В исходной таблице 'data.csv' пропуски наблюдались в колонках 'total_income' и 'days_employed'. \
Пропуски могут быть вызваны множеством причин: ошибка при вводе данных, потеря данных при импорте/экспорте и др. Некоторые пропуски могут быть оставлены специально. \
Можно предположить, что в столбце 'total_income' и 'days_employed' пропуски появились из-за записи дробных чисел через запятую, как это принято в некоторых странах, либо это ошибки при вычислениях (например, деление на 0).

### Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 

На медиану, в отличие от среднего арифметического, не влияют крайности – необычно большое или необычно маленькое значение.

## Шаг 4: общий вывод.

На факт погашения кредита в срок больше всего влияет наличие и количество детей (см. п.3.1)\
Наиболее надежные клиенты - бездетные (7.5% невозврата)\
Безответственные - с 4 детьми (9.8% невозврата)

Семейное положение в совокупности с возрастом также оказалось фактором погашения кредита (см. п.3.2) \
Наиболее надежные клиенты - возрастные вдовцы / вдовы (6.6% невозврата)\
Безответственные - молодые не женатые (9.8% невозврата)

Уровень дохода также показал связь с возвратом кредита в срок (см. п.3.3)\
Наиболее надежные клиенты - с высоким уровнем дохода (7.1%)\
Наименее надежные - с низким уровнем дохода (8.5%)

Также была обнаружена зависимость между целью кредита и его возвратом. (см. п.3.4)\
Наиболее ответственные – приобретатели недвижимости (7.3% невозврата)\
Наименее ответственные - кредиторы на образование и автомобили (9.3% невозврата)

Наш идеальный клиент: бездетный вдовец с высоким достатком, берущий кредит на квартиру.\
Проблемный клиент: молодой многодетный холостяк с низким достатком, берущий кредит на машину.